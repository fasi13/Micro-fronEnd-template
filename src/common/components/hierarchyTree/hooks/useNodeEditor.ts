import { useRef, useState } from 'react';
import { ErrorResponse } from '../../../../types';

export interface useNodeEditorProps {
	data: string | undefined;
	isApplication: boolean;
	setError: (val: ErrorResponse | string) => void;
	clearError: () => void;
	onSubmit: (value: string) => void;
}

interface useNodeEditorReturnType {
	checkValidityAndSubmit: () => void;
	preValue: React.MutableRefObject<string>;
	value: string;
	nodeEditorPlaceHolder: () => string;
	setValue: React.Dispatch<React.SetStateAction<string>>;
}

export const useNodeEditor = (
	props: useNodeEditorProps,
): useNodeEditorReturnType => {
	const { data, isApplication, setError, clearError, onSubmit } = props;

	const [value, setValue] = useState(data || '');
	const preValue = useRef(value);

	const validAppPattern = /^[^)(]+(\([^()]+\)){1}$/g;
	const emptyErrorMsg = 'Value can not be empty';
	const invalidApplicationValueFormatMsg =
		'Application format should be: Application Name (Value)';

	const checkValidityAndSubmit = (): void => {
		if (value === '') {
			setError(emptyErrorMsg);
		} else if (
			isApplication &&
			value.match(validAppPattern)?.filter(m => m !== undefined).length !== 1
		) {
			setError(invalidApplicationValueFormatMsg);
		} else if (preValue.current !== value) {
			clearError();
			preValue.current = value;
			onSubmit(value);
		}
	};

	const nodeEditorPlaceHolder = (): string => {
		if (isApplication) return 'Add New Application';
		return 'Add New Application Group';
	};

	return {
		checkValidityAndSubmit,
		nodeEditorPlaceHolder,
		preValue,
		value,
		setValue,
	};
};
