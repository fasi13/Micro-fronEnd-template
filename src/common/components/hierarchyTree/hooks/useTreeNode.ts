// import { useCallback, useEffect, useReducer } from 'react';
// import { useBreadcrumbStore, useHierarchyStore } from '../../../../state';
// import {
// 	ErrorResponse,
// 	NodeActions,
// 	NodePath,
// 	TreeView,
// } from '../../../../types';

// type TEditor = '' | 'Application' | 'Group';
// type ActionType =
// 	| { type: 'LOAD_CHILDREN' }
// 	| { type: 'LOAD_CHILDREN_SUCCESS' }
// 	| { type: 'LOAD_CHILDREN_ERROR'; val: ErrorResponse }
// 	| { type: 'TOGGLE_NEW_EDITOR'; val: TEditor }
// 	| { type: 'TOGGLE_EDIT'; val: boolean }
// 	| { type: 'SAVE'; val: string }
// 	| { type: 'SAVING_SUCCESS' }
// 	| { type: 'SAVING_ERROR'; val: ErrorResponse }
// 	| { type: 'SAVE_RESET' }
// 	| { type: 'SET_ERROR'; val: ErrorResponse | string }
// 	| { type: 'CLEAR_ERROR' }
// 	| { type: 'COLLAPSE_CHILDREN_NODES' };

// interface TNodeDetail extends NodeActions {
// 	data: TreeView;
// 	nodeId: number;
// 	nodePath: NodePath[];
// 	edit: boolean | null;
// 	saving: boolean;
// 	toggleNewEditor: TEditor;
// 	newValue: string | null;
// 	error: string | null;
// 	loadingChildren: boolean;
// 	isFetched: boolean;
// 	nodeType: 'Application' | 'Group';
// }
// export interface TResponse {
// 	nodeState: TNodeDetail;
// 	toggleNewEditor: (val: TEditor) => void;
// 	editNode: (val: boolean) => void;
// 	toggleChildren: () => void;
// 	save: (value: string) => void;
// 	setError: (val: ErrorResponse | string) => void;
// 	clearError: () => void;
// }

// const extractApplicationNameAndValue = (text: string): string[] => {
// 	const newLocal = /^[^)(]+|\([^)(]*\)$/g;
// 	return text.match(new RegExp(newLocal))?.map(x => x.toString()) || ['', ''];
// };

// const reducer = (prevState: TNodeDetail, action: ActionType): TNodeDetail => {
// 	switch (action.type) {
// 		case 'LOAD_CHILDREN': {
// 			return {
// 				...prevState,
// 				loadingChildren: true,
// 				toggleNewEditor: '',
// 			};
// 		}
// 		case 'LOAD_CHILDREN_SUCCESS': {
// 			return { ...prevState, loadingChildren: false, isFetched: true };
// 		}
// 		case 'LOAD_CHILDREN_ERROR': {
// 			return {
// 				...prevState,
// 				loadingChildren: false,
// 				error: action.val.errors?.[0],
// 				isFetched: false,
// 			};
// 		}
// 		case 'TOGGLE_EDIT':
// 			return { ...prevState, edit: action.val, toggleNewEditor: '' };
// 		case 'TOGGLE_NEW_EDITOR': {
// 			if (action.val) return { ...prevState, toggleNewEditor: action.val };

// 			return {
// 				...prevState,
// 				toggleNewEditor: action.val,
// 				error: null,
// 			};
// 		}
// 		case 'SAVE': {
// 			return { ...prevState, newValue: action.val, saving: true, error: null };
// 		}
// 		case 'SAVING_SUCCESS':
// 			return {
// 				...prevState,
// 				saving: false,
// 				toggleNewEditor: '',
// 				loadingChildren: false,
// 				error: null,
// 				isFetched: false,
// 			};
// 		case 'SAVING_ERROR':
// 			return { ...prevState, saving: false, error: action.val.errors?.[0] };
// 		case 'SAVE_RESET':
// 			return {
// 				...prevState,
// 				saving: false,
// 				toggleNewEditor: '',
// 				loadingChildren: false,
// 				error: null,
// 			};
// 		case 'SET_ERROR': {
// 			if (typeof action.val === 'string')
// 				return { ...prevState, error: action.val };
// 			return { ...prevState, error: action.val.errors?.[0] };
// 		}
// 		case 'CLEAR_ERROR':
// 			return { ...prevState, error: null };
// 		default:
// 			throw new Error();
// 	}
// };

// export const useTreeNode = (nodeDetail: TNodeDetail): TResponse => {
// 	const [nodeState, dispatch] = useReducer(reducer, nodeDetail);
// 	const { setBreadCrumb } = useBreadcrumbStore();

// 	const toggleNewEditor = useCallback((val: TEditor) => {
// 		dispatch({ type: 'TOGGLE_NEW_EDITOR', val });
// 	}, []);

// 	const editNode = useCallback((val: boolean) => {
// 		dispatch({ type: 'TOGGLE_EDIT', val });
// 	}, []);

// 	const save = useCallback((value: string) => {
// 		dispatch({ type: 'SAVE', val: value });
// 	}, []);

// 	const setError = useCallback((val: ErrorResponse | string) => {
// 		dispatch({ type: 'SET_ERROR', val });
// 	}, []);

// 	const clearError = useCallback(() => {
// 		dispatch({ type: 'CLEAR_ERROR' });
// 	}, []);

// 	const savingCb = useCallback((err: ErrorResponse | null) => {
// 		if (err) {
// 			dispatch({ type: 'SAVING_ERROR', val: err });
// 		} else if (nodeState.edit === false) {
// 			// refetch
// 			dispatch({ type: 'SAVING_SUCCESS' });
// 			// call  state method to collapse children
// 			// store.collapseChildren(node);
// 		}
// 	}, []);

// 	const loadNodeChildren = useCallback(() => {
// 		const { data, nodeId, nodePath, getChildren, onToggleCollapse } =
// 			nodeDetail;

// 		getChildren(data, nodeId, nodePath, (err: ErrorResponse | null) => {
// 			if (err) dispatch({ type: 'LOAD_CHILDREN_ERROR', val: err });
// 			else {
// 				dispatch({ type: 'LOAD_CHILDREN_SUCCESS' });
// 				onToggleCollapse(nodePath, false);
// 				setBreadCrumb(nodePath);
// 			}
// 		});
// 	}, [nodeState.loadingChildren]);

// 	const editApplication = useCallback(
// 		(name: string, value: string) => {
// 			if (nodeState.saving) {
// 				const { data, nodeId, nodePath, onEditApplication } = nodeDetail;
// 				onEditApplication(
// 					data,
// 					nodeId,
// 					nodePath,
// 					name,
// 					value.replaceAll(/\(|\)/gi, ''),
// 					savingCb,
// 				);
// 			}
// 		},
// 		[nodeState.saving],
// 	);

// 	const editGroup = useCallback(
// 		(name: string) => {
// 			if (nodeState.saving) {
// 				const { data, nodeId, nodePath, onEditGroup } = nodeDetail;
// 				onEditGroup(data, nodeId, nodePath, name, savingCb);
// 			}
// 		},
// 		[nodeState.saving],
// 	);

// 	const addApplication = useCallback(
// 		(name: string, value: string) => {
// 			if (nodeState.saving) {
// 				const { data, nodeId, nodePath, onAddApplication } = nodeDetail;
// 				onAddApplication(
// 					data,
// 					nodeId,
// 					nodePath,
// 					name,
// 					value.replaceAll(/\(|\)/gi, ''),
// 					savingCb,
// 				);
// 			}
// 		},
// 		[nodeState.saving],
// 	);

// 	const addGroup = useCallback(
// 		(name: string) => {
// 			if (nodeState.saving) {
// 				const { data, nodeId, nodePath, onAddGroup } = nodeDetail;
// 				onAddGroup(data, nodeId, nodePath, name, savingCb);
// 			}
// 		},
// 		[nodeState.saving],
// 	);

// 	const toggleChildren = useCallback(() => {
// 		const { data, nodePath, onToggleCollapse } = nodeDetail;

// 		if (data.collapsed && !nodeState.isFetched && !data.childrenData)
// 			dispatch({ type: 'LOAD_CHILDREN' });
// 		else if (data.collapsed && nodeState.isFetched && data.childrenData)
// 			onToggleCollapse(nodePath, false);
// 		else if (!data.collapsed) {
// 			// just collapse
// 			onToggleCollapse(nodePath, true);
// 		}
// 	}, [nodeDetail]);

// 	useEffect(() => {
// 		if (nodeState.saving) {
// 			const { newValue } = nodeState;
// 			if (nodeState.edit) {
// 				if (nodeState.nodeType === 'Application' && newValue) {
// 					const [name, value] = extractApplicationNameAndValue(newValue);
// 					editApplication(name, value);
// 				} else if (nodeState.nodeType === 'Group' && newValue) {
// 					editGroup(newValue);
// 				}
// 			} else if (nodeState.toggleNewEditor === 'Application' && newValue) {
// 				const [name, value] = extractApplicationNameAndValue(newValue);
// 				addApplication(name, value);
// 			} else if (nodeState.toggleNewEditor === 'Group' && newValue) {
// 				addGroup(newValue);
// 			}
// 		}
// 	}, [nodeState.saving]);

// 	useEffect(() => {
// 		if (nodeState.loadingChildren) loadNodeChildren();
// 	}, [nodeState.loadingChildren]);

// 	useEffect(() => {
// 		if (!nodeState.edit) dispatch({ type: 'CLEAR_ERROR' });
// 	}, [nodeState.edit]);

// 	useEffect(() => {
// 		console.log('???? -->', nodeDetail.data);
// 	}, []);

// 	return {
// 		nodeState,
// 		toggleNewEditor,
// 		toggleChildren,
// 		editNode,
// 		save,
// 		setError,
// 		clearError,
// 	};
// };


export { };
