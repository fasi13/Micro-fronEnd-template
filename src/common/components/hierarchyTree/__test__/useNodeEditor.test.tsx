import { renderHook } from '@testing-library/react-hooks';
import { useNodeEditor, useNodeEditorProps } from '../hooks/useNodeEditor';

describe('useNodeEditor', () => {
	let dummyProps: useNodeEditorProps;

	beforeEach(() => {
		dummyProps = {
			data: '',
			error: '',
			isApplication: true,
			setError: jest.fn(),
			clearError: jest.fn(),
			onSubmit: jest.fn(),
		};
	});
	describe('useNodeEditor snapShot return values', () => {
		it('returns object of helper methods', () => {
			const result = renderHook(useNodeEditor, { initialProps: dummyProps });
			expect(result).toMatchSnapshot();
		});
	});

	describe('nodeEditorPlaceHolder', () => {
		it(`returns 'Add New Application' if node is Application`, () => {
			dummyProps.isApplication = true;
			const { result } = renderHook(useNodeEditor, {
				initialProps: dummyProps,
			});
			expect(result.current.nodeEditorPlaceHolder()).toMatch(
				'Add New Application',
			);
		});

		it(`returns 'Add New Application Group' if node is not Application`, () => {
			dummyProps.isApplication = false;
			const { result } = renderHook(useNodeEditor, {
				initialProps: dummyProps,
			});
			expect(result.current.nodeEditorPlaceHolder()).toMatch(
				'Add New Application Group',
			);
		});
	});

	describe('checkValidityAndSubmit', () => {
		it('it calls set Error if passed in value is empty string', () => {
			dummyProps.data = '';

			const { result } = renderHook(useNodeEditor, {
				initialProps: dummyProps,
			});
			result.current.checkValidityAndSubmit();

			expect(dummyProps.setError).toHaveBeenCalledWith(
				'Value can not be empty',
			);
		});

		it(`checks if value has valid pattern name (value) if isApplication is true ,  if pattern is wrong setError gets called `, () => {
			dummyProps.data = 'name ((value))'; // wrong pattern
			dummyProps.isApplication = true;

			const { result } = renderHook(useNodeEditor, {
				initialProps: dummyProps,
			});
			result.current.checkValidityAndSubmit();

			expect(dummyProps.setError).toHaveBeenCalledWith(
				'Application format should be: Application Name (Value)',
			);
		});
		it('checks if the input form is submit properly', () => {
			dummyProps.data = 'New Application Group(2021)'; // wrong pattern
			dummyProps.isApplication = true;
			const { result } = renderHook(useNodeEditor, {
				initialProps: dummyProps,
			});
			result.current.preValue.current = 'New Application Group(2026)';
			result.current.checkValidityAndSubmit();
			expect(dummyProps.onSubmit).toBeCalledTimes(1);
		});

		it(`clears editor value if error existed and preValue is not equal to current Editor Value`, () => {
			dummyProps.error = 'Value can not be empty';

			const { result } = renderHook(useNodeEditor, {
				initialProps: dummyProps,
			});

			result.current.preValue.current = 'old value';
			const mockChangeEvent = {
				currentTarget: {
					value: '',
				},
				target: {
					value: 'new value',
				},
			} as React.ChangeEvent<HTMLInputElement>;

			result.current.setEditorValue(mockChangeEvent);
			expect(dummyProps.clearError).toHaveBeenCalled();
		});

		it('sets the value of the editor', () => {
			const { result } = renderHook(useNodeEditor, {
				initialProps: dummyProps,
			});
			const mockChangeEvent = {
				currentTarget: {
					value: '',
				},
				target: {
					value: 'new value',
				},
			} as React.ChangeEvent<HTMLInputElement>;

			result.current.setEditorValue(mockChangeEvent);

			expect(result.current.value).toMatch('new value');
		});
	});
});
