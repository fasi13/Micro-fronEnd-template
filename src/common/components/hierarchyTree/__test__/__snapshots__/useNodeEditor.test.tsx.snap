// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useNodeEditor useNodeEditor snapShot return values returns object of helper methods 1`] = `
Object {
  "rerender": [Function],
  "result": Object {
    "all": Array [
      Object {
        "checkValidityAndSubmit": [Function],
        "closeButtonStyling": [Function],
        "nodeEditorPlaceHolder": [Function],
        "preValue": Object {
          "current": "",
        },
        "setEditorValue": [Function],
        "value": "",
      },
    ],
    "current": Object {
      "checkValidityAndSubmit": [Function],
      "closeButtonStyling": [Function],
      "nodeEditorPlaceHolder": [Function],
      "preValue": Object {
        "current": "",
      },
      "setEditorValue": [Function],
      "value": "",
    },
    "error": undefined,
  },
  "unmount": [Function],
  "waitFor": [Function],
  "waitForNextUpdate": [Function],
  "waitForValueToChange": [Function],
}
`;
